modules/library/main/src/main/java/org/geotools/geometry/jts/JTS.java

```diff
@@ -1685,23 +1688,61 @@ public static CoordinateReferenceSystem getCRS(Geometry geometry) {
             if (crs != null) {
                 return crs;
             }
+        } else if (geometry.getSRID() > 0) {
+            try {
+                return CRS.decode("EPSG:" + geometry.getSRID());
+            } catch (Exception e) {
+                return null;
+            }
         }
         return null;
     }
 
     /**
-     * Set the CRS of a geometry.
+     * Set the CRS of a geometry is {@link Geometry#getUserData()}.
+     *
+     * <p>Stored in {@link Geometry#getUserData()} as a {@code CoordinateReferenceSystem} directly,
+     * or as a {@code Map<Class,Value>}.
+     *
+     * <p>Stored in {@link Geometry#getSRID()} as EPSG code.
      *
      * @param geometry the geometry
      * @param crs the CRS
      */
     public static void setCRS(Geometry geometry, CoordinateReferenceSystem crs) {
-        Object obj = geometry.getUserData();
-        Map<Object, Object> userData = new HashMap<>();
-        if (obj != null && obj instanceof Map) {
-            userData.putAll((Map<?, ?>) obj);
+        if (crs != null) {
+            try {
+                Integer srid = CRS.lookupEpsgCode(crs, false);
+                if (srid != null) {
+                    geometry.setSRID(srid);
+                }
+            } catch (FactoryException noSRID) {
+                geometry.setSRID(0);
+            }
+
+            Object obj = geometry.getUserData();
+            if (obj != null && obj instanceof CoordinateReferenceSystem) {
+                geometry.setUserData(crs);
+            }
+            Map<Object, Object> userData = new HashMap<>();
+            if (obj != null && obj instanceof Map) {
+                // preserve existing user data map contents
+                userData.putAll((Map<?, ?>) obj);
```

Comment:
@aaime in the interests of doing less working in possibly time sensitive method, is there any reason I should not update the existing user data map in place?

Reply:
I was wondering the same, but I cannot offer an answer. I'd suggest checking the history ... @NielsCharlier introduced that method, maybe he can explain the reasoning (maybe it's just basic defensive programming).

Reply:
Yeah I already asked @NielsCharlier for input in an earlier message. I am going to try the change locally and let it build over night.